chainer/link_hook.py

```diff
@@ -0,0 +1,139 @@
+        with :class:`~chainer.link_hooks.TimerHook`, which is a subclass of
+        :class:`~chainer.LinkHook`.
+
+        >>> from chainer import link_hooks
+        >>> class Model(chainer.Chain):
+        ...   def __init__(self):
+        ...     super(Model, self).__init__()
+        ...     with self.init_scope():
+        ...       self.l = L.Linear(10, 10)
+        ...   def forward(self, x1):
+        ...     return F.exp(self.l(x1))
+        >>> model1 = Model()
+        >>> model2 = Model()
+        >>> x = chainer.Variable(np.zeros((1, 10), np.float32))
+        >>> with chainer.function_hooks.TimerHook() as m:
+        ...   _ = model1(x)
+        ...   y = model2(x)
+        ...   print("Total time : " + str(m.total_time()))
+        ...   model3 = Model()
+        ...   z = model3(y) # doctest:+ELLIPSIS
+        Total time : ...
+
+        In this example, we measure the elapsed times for each forward
+        propagation of all functions in ``model1`` and ``model2``
+        (specifically, :class:`~chainer.functions.LinearFunction` and
+        :class:`~chainer.functions.Exp` of ``model1`` and ``model2``).
+        Note that ``model3`` is not a target of measurement
+        as :class:`~chainer.function_hooks.TimerHook` is unregistered
+        before forward propagation of ``model3``.
+
+    .. note::
+
+       Chainer stores the dictionary of registered function hooks
+       as a thread local object. So, function hooks registered
+       are different depending on threads.
+
+    The other one is to register directly to
+    :class:`~chainer.FunctionNode` object with
+    :meth:`~chainer.Function.add_hook` method.
+    Function hooks registered in this way can be removed by
+    :meth:`~chainer.Function.delete_hook` method.
+    Contrary to former registration method, function hooks are registered
+    only to the function which :meth:`~chainer.FunctionNode.add_hook`
+    is called.
+
+    Args:
+        name(str): Name of this function hook.
+    """
+
+    name = 'LinkHook'
+
+    def __enter__(self):
+        link_hooks = chainer.get_link_hooks()
+        if self.name in link_hooks:
+            raise KeyError('hook %s already exists' % self.name)
+
+        link_hooks[self.name] = self
+        self.added()
+        return self
+
+    def __exit__(self, *_):
+        chainer.get_link_hooks()[self.name].deleted()
+        del chainer.get_link_hooks()[self.name]
+
+    def added(self, link=None):
+        """Callback function invoked when a link hook is added
+
+        Args:
+            link(~chainer.Link): Link object to which
+                the link hook is added.
+        """
+        pass
+
+    def deleted(self, link=None):
+        """Callback function invoked when a link hook is deleted
+
+        Args:
+            link(~chainer.Link): Link object to which
+                the link hook is deleted.
+        """
+        pass
+
+    # forward
+    def forward_preprocess(self, link, *args, **kwargs):
```

Comment:
It shouldn't be limited to `forward`.

Reply:
I re-checked https://github.com/chainer/chainer/issues/2748#issuecomment-383836796, and it seems we had the intention that arbitrary user-defined methods (other than `forward`) can also be considered as a kind of "forward" methods.